File: client-api/src/main/java/org/ostelco/topup/api/resources/ProfileResource.java
package org.ostelco.topup.api.resources;

import io.dropwizard.auth.Auth;
import io.vavr.control.Either;
import org.ostelco.prime.model.Subscriber;
import org.ostelco.topup.api.auth.AccessTokenPrincipal;
import org.ostelco.topup.api.core.Error;
import org.ostelco.topup.api.db.SubscriberDAO;

import javax.validation.constraints.NotNull;
import javax.ws.rs.Consumes;
import javax.ws.rs.GET;
import javax.ws.rs.POST;
import javax.ws.rs.PUT;
import javax.ws.rs.Path;
import javax.ws.rs.Produces;
import javax.ws.rs.core.Response;

/**
 * Profile API.
 *
 */
@Path("/profile")
public class ProfileResource extends ResourceHelpers {

    private final SubscriberDAO dao;

    public ProfileResource(SubscriberDAO dao) {
        this.dao = dao;
    }

    @GET
    @Produces({"application/json"})
    public Response getProfile(@Auth AccessTokenPrincipal token) {
        if (token == null) {
            return Response.status(Response.Status.UNAUTHORIZED)
                .build();
        }

        Either<Error, Subscriber> result = dao.getProfile(token.getName());

        return result.isRight()
            ? Response.status(Response.Status.OK)
                 .entity(asJson(result.right().get()))
                 .build()
            : Response.status(Response.Status.NOT_FOUND)
                 .build();
    }

    @POST
    @Consumes({"application/json"})
    public Response createProfile(@Auth AccessTokenPrincipal token,
            @NotNull final Subscriber profile) {
        if (token == null) {
            return Response.status(Response.Status.UNAUTHORIZED)
                .build();
        }

        Either<Error, Subscriber> result = dao.createProfile(token.getName(), profile);

        return result.isRight()
                ? Response.status(Response.Status.CREATED)
                .entity(asJson(result.right().get()))
                .build()
                : Response.status(Response.Status.FORBIDDEN)
                .build();
    }

    @PUT
    @Consumes({"application/json"})
    public Response updateProfile(@Auth AccessTokenPrincipal token,
            @NotNull final Subscriber profile) {
        if (token == null) {
            return Response.status(Response.Status.UNAUTHORIZED)
                .build();
        }

        Either<Error, Subscriber> result = dao.updateProfile(token.getName(), profile);

        return result.isRight()
                ? Response.status(Response.Status.OK)
                .entity(asJson(result.right().get()))
                .build()
                : Response.status(Response.Status.NOT_FOUND)
                .build();
    }
}

File: client-api/src/main/kotlin/org/ostelco/prime/client/api/resources/ProfileResource.kt
package org.ostelco.prime.client.api.resources

import io.dropwizard.auth.Auth
import org.ostelco.prime.auth.AccessTokenPrincipal
import org.ostelco.prime.client.api.store.SubscriberDAO
import org.ostelco.prime.jsonmapper.asJson
import org.ostelco.prime.model.Customer
import org.ostelco.prime.model.Identity
import java.util.*
import javax.validation.constraints.NotNull
import javax.ws.rs.Consumes
import javax.ws.rs.GET
import javax.ws.rs.POST
import javax.ws.rs.PUT
import javax.ws.rs.Path
import javax.ws.rs.Produces
import javax.ws.rs.QueryParam
import javax.ws.rs.core.MediaType
import javax.ws.rs.core.Response

/**
 * Profile API.
 *
 */
@Path("/profile")
class ProfileResource(private val dao: SubscriberDAO) {

    @GET
    @Produces(MediaType.APPLICATION_JSON)
    fun getProfile(@Auth token: AccessTokenPrincipal?): Response {
        if (token == null) {
            return Response.status(Response.Status.UNAUTHORIZED)
                    .build()
        }

        return dao.getProfile(identity = Identity(id = token.name, type = "EMAIL", provider = token.provider))
                .fold(
                        { apiError -> Response.status(apiError.status).entity(asJson(apiError)) },
                        { Response.status(Response.Status.OK).entity(asJson(it)) })
                .build()
    }

    @POST
    @Produces(MediaType.APPLICATION_JSON)
    @Consumes(MediaType.APPLICATION_JSON)
    fun createProfile(@Auth token: AccessTokenPrincipal?,
                      @NotNull profile: Customer,
                      @QueryParam("referred_by") referredBy: String?): Response {

        if (token == null) {
            return Response.status(Response.Status.UNAUTHORIZED)
                    .build()
        }

        return dao.createProfile(
                identity = Identity(id = token.name, type = "EMAIL", provider = token.provider),
                profile = profile.copy(
                        id = UUID.randomUUID().toString(),
                        email = token.name,
                        referralId = UUID.randomUUID().toString()),
                referredBy = referredBy)
                .fold(
                        { apiError -> Response.status(apiError.status).entity(asJson(apiError)) },
                        { Response.status(Response.Status.CREATED).entity(asJson(it)) })
                .build()
    }

    @PUT
    @Produces(MediaType.APPLICATION_JSON)
    @Consumes(MediaType.APPLICATION_JSON)
    fun updateProfile(@Auth token: AccessTokenPrincipal?,
                      @NotNull profile: Customer): Response {
        if (token == null) {
            return Response.status(Response.Status.UNAUTHORIZED)
                    .build()
        }

        return dao.updateProfile(
                identity = Identity(id = token.name, type = "EMAIL", provider = token.provider),
                profile = profile)
                .fold(
                        { apiError -> Response.status(apiError.status).entity(asJson(apiError)) },
                        { Response.status(Response.Status.OK).entity(asJson(it)) })
                .build()
    }
}
